### PR TITLE
ldd: Ignore symlinks that lead to the same file

### PR DESCRIPTION
**Pull request by Foxboron**
Usually virtual libraries like `vdso` is printed on the same line,
howeve this doesn't seem to always be the case.

Implement a check to see if the symlink is pointing at the same file and
ignore the entry if it does.

We also return an error instead of failing with a printed message.

Signed-off-by: Morten Linderud <morten.linderud@mullvad.net>

### CODE REVIEW

**Comment by Foxboron** [comment]
An alternative to this patch would be to hook up `SkipLDD` properly as a flag. But I'm not sure if it's a better solution than ignoring the possibly missing so-names and issue warnings.

```
diff --git a/u-root.go b/u-root.go
index cefb8a002..cf2be2c6b 100644
--- a/u-root.go
+++ b/u-root.go
@@ -49,6 +49,7 @@ var (
 	statsOutputPath                         *string
 	statsLabel                              *string
 	shellbang                               *bool
+	skipLDD                                 *bool
 	tags                                    *string
 	// For the new gobusybox support
 	usegobusybox *bool
@@ -85,6 +86,8 @@ func init() {

 	shellbang = flag.Bool("shellbang", false, "Use #! instead of symlinks for busybox")

+	skipLDD = flag.Bool("skip-ldd", false, "Skip ldd resolution")
+
 	statsOutputPath = flag.String("stats-output-path", "", "Write build stats to this file (JSON)")
 	statsLabel = flag.String("stats-label", "", "Use this statsLabel when writing stats")

@@ -356,6 +359,7 @@ func Main(l ulog.Logger, buildOpts *gbbgolang.BuildOpts) error {
 		UseExistingInit: *useExistingInit,
 		InitCmd:         initCommand,
 		DefaultShell:    *defaultShell,
+		SkipLDD:         *skipLDD,
 		BuildOpts:       buildOpts,
 	}
 	uinitArgs := shlex.Argv(*uinitCmd)
```

**Review by rminnich — COMMENTED** [review]
There's an even worse problem here, that a package is calling log.Fatal anyway, that may have happened when bits of ldd were turned into a package, but it's a mistake.
At best it should have returned an error but you are right, blowing out on not being able to follow something seems broken.

Is the vdso the only such "non existant file" and should we explicitly skip it if the prefix of the library is linux-vdso?

The weird thing is, when I use the lddfiles command, which uses this package, I never see the error. Had you run into a problem in real life?

**Comment by Foxboron** [comment]
>Is the vdso the only such "non existant file" and should we explicitly skip it if the prefix of the library is linux-vdso?

We could do it this way, that was my original thought. But since we have `log.Fatal` I thought the first iteration could atleast soften this.

>The weird thing is, when I use the lddfiles command, which uses this package, I never see the error. Had you run into a problem in real life?

Yes.

On Arch the modules directories has several binaries which contains this file. Just a simple file inclusion would give a hard error.

`./u-root -files "/usr/lib/modules/5.19.1-arch2-1" -format=cpio`

**Comment by 10000TB** [comment]
+1 returning the error 

Imagine calling this as a best effort, in some program, and then the program exited :P (in whichever rare case that may happen)

**Comment by Foxboron** [comment]
Okay, if we return the error then we need to deal with these fake .so-name dependencies. We can just ignore them I guess and return the error?

**Comment by rminnich** [comment]
OK, so two things
the log.Printf should return an error, not print.

Secondly, arch linux somehow managed to really break things.

Here is ldd on pretty much every linux I've ever used (debian, ubuntu, fedora tested just now)
```
	linux-vdso.so.1 (0x00007fffc309d000)
```
tinycore:
```
ldd /bin/date
	linux-gate.so.1 (0xf7f4e000)
	libcrypt.so.1 => /lib/libcrypt.so.1 (0xf7f16000)
	libm.so.6 => /lib/libm.so.6 (0xf7e53000)
	libc.so.6 => /lib/libc.so.6 (0xf7d06000)
	/lib/ld-linux.so.2 (0xf7f50000)
```
Here is arch:
	linux-vdso.so.1 => linux-vdso.so.1 (0x00007ffdd55ed000)

So the usual notation for this output is: => implies a symlink to a file. That's been that way for decades. Arch screwed up.

Hence the docs in the package.
```
// runinterp runs the interpreter with the --list switch
// and the file as an argument. For each returned line
// it looks for => as the second field, indicating a
// real .so (as opposed to the .vdso or a string like
// 'not a dynamic executable'.
```
 So the question is, given that arch got it wrong, what do we do in this package?

**Comment by Foxboron** [comment]
I missed the last comment so ignore the pushed changes if you review them.

I'm not sure if this is "Arch screwing up" or if this is a libc change will check. But very weird.

**Comment by rminnich** [comment]
It *seems* that we could check for this case:
linux-vdso.so.1 => linux-vdso.so.1
by just checking if it is a symlink to itself, and ignoring it if so.

**Comment by Foxboron** [comment]
I'm still curious why this diffes across Arch and otherr distros.

From reading the source it seems like this isn't always expected, but it's very weird that the filename is the same yet the condition fails.

https://sourceware.org/git/?p=glibc.git;a=blob;f=elf/rtld.c;h=3e771a93d83457665d0e143cf6c22ace592ec6ac;hb=HEAD#l2126

I'll do a take on this where we check for the symlink.

**Comment by Foxboron** [comment]
This iteration does the suggested approach. Just had to trim a bit of spaces. The test suite passes on my Arch machine.

**Review by rminnich — APPROVED** [review]

File: pkg/ldd/ldd_unix.go [hunk 1]
```diff
@@ -80,13 +79,16 @@ func runinterp(interp, file string) ([]string, error) {
 		return nil, err
 	}
 	for _, p := range strings.Split(string(o), "\n") {
-		f := strings.Split(p, " ")
```
Comment (rminnich) [inline pkg/ldd/ldd_unix.go:83]: If you turn the core of this loop into a function, writing a test will be very easy, I hope the extra headache of doing that is ok.

File: pkg/ldd/ldd_unix.go [hunk 2]
```diff
@@ -80,13 +79,16 @@ func runinterp(interp, file string) ([]string, error) {
...
+		f := strings.Split(strings.TrimSpace(p), " ")
```
Comment (abrender) [inline pkg/ldd/ldd_unix.go:82]: Will strings.Fields(strings.TrimSpace(p)) work here (https://pkg.go.dev/strings#Fields)?  That will split by "one or more consecutive white space characters"

Comment (Foxboron) [reply pkg/ldd/ldd_unix.go:82]: Oh, that is neat. Never seen that function before.

**Comment by Foxboron** [comment]
Implemented some tests and split the parsing part of `runinterp` into `parseinterp`

**Pull request merged**